File: strings/longest_palindrome_substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        if s == s[::-1]: return s

        window_len = 1
        start = 0
        for i in range(1, len(s)):
            if i - window_len >= 0 and s[i - window_len:i+1] == s[i - window_len:i+1][::-1]:
                start = i - window_len
                window_len += 1

            elif i - window_len >= 1 and s[i - window_len - 1:i + 1] == s[i - window_len - 1:i + 1][::-1]:
                start = i - window_len - 1
                window_len += 2

        return s[start: start + window_len]

File: strings/longest_palindrome_substring.rb
# @param {String} s
# @return {String}
def longest_palindrome(s)
    return s if s == s.reverse

    window_size = 1
    start = 0
    
    (1...s.size).each do |i|
        if i - window_size >= 0 && s[i - window_size..i] == s[i - window_size..i].reverse
            start = i - window_size
            window_size += 1
        elsif i - window_size >= 1 && s[i - window_size - 1..i] == s[i - window_size - 1..i].reverse
            start = i - window_size - 1
            window_size += 2
        end
    end
    s[start, window_size]
end